src/charm.py

```diff
@@ -1,104 +1,161 @@
+        )
+
+    @property
+    def has_db(self) -> bool:
+        """Only consider a DB connection if we have config info."""
+        rel = self.model.get_relation(DATABASE)
+        return len(rel.units) > 0 if rel is not None else False
+
+    @property
+    def peers(self) -> list:
+        """Fetch the peer relation."""
+        return self.model.get_relation(PEER)
+
+    def set_peer_data(self, key: str, data: Any) -> None:
+        """Put information into the peer data."""
+        if self.unit.is_leader():
+            self.peers.data[self.app][key] = json.dumps(data)
+
+    def get_peer_data(self, key: str) -> Any:
+        """Retrieve information from the peer data."""
+        data = self.peers.data[self.app].get(key, "{}")
+        return json.loads(data)
+
+    def _configure(self) -> None:
+        """Configure the charm."""
+        # Get the relation data
+        data = self.get_peer_data(DATABASE)
+
+        config_data = json.loads(data["mysql"])
+
+        # Define Pebble layer configuration
+        pebble_layer = self._mysqlrouter_layer(
+            port=config_data["port"],
+            host=config_data["host"],
+            user=config_data["user"],
+            password=config_data["password"],
+        )
         # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
-        container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ActiveStatus()
+        container = self.unit.get_container(self.name)
+        plan = container.get_plan()
 
-    def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
+        if plan.services != pebble_layer.services:
+            logger.info("Config changed")
+            container.add_layer(self.name, pebble_layer, combine=True)
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
+            container.restart(self.name)
+            logging.info("mysqlrouter restarted")
 
-        Learn more about config at https://juju.is/docs/sdk/config
-        """
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
+        self.unit.status = ActiveStatus()
 
-    def _on_fortune_action(self, event):
-        """Just an example to show how to receive actions.
+    def _on_mysqlrouter_pebble_ready(self, event) -> None:
+        """Define and start a workload using the Pebble API."""
+        if not self.has_db:
```

Comment:
is there a race condition where `self.has_db` is `true`, but the `_on_database_relation_created` handler has not fired yet and thus has not `set_peer_data`? if this happens, then `_configure` will retrieve empty host/port/user/password

may be we can add a TODO if this race condition exists to handle the case properly in the interest of keeping this PR short